rel/expr_tuple.go

```diff
@@ -151,3 +155,22 @@ func (e *TupleExpr) Eval(ctx context.Context, local Scope) (Value, error) {
 	// TODO: Construct new tuple directly
 	return tuple.(*GenericTuple).Canonical(), nil
 }
+
+func (e *TupleExpr) evalTests(ctx context.Context, local Scope) (Value, error) {
```

Comment:
As above, this should be in `internal/test`. It could probably be an extension of the `isRecursivelyTrue` function, rather than a separate thing (i.e. in a test, we would want to report all of the errors at all levels of a data structure, not just for tuples and not just one level deep).

Reply:
Current code can't support this approach. The code `value, err := expr.Eval(ctx, scope)` calculates the value, as its implementation when one error happens in evaluation, it will stop executing and return a `nil` as `value`. So it can't calculate the value with some code like `isRecursivelyTrue`.
And as current `expr` definition, it can't do the similar thing like `isRecursivelyTrue` too, refers https://github.com/arr-ai/arrai/pull/631#discussion_r487479789.
I think it has to do a leverage, open access to attributes in `expr` or add some functions in package `rel` which can be used by `test`.

Reply:
So my thinking is:

1. The `_test.arrai` file should at least compile. If it doesn't you shouldn't expect good error reporting.
2. If the test compiles, then its output will be a data structure: a set, array, dict or tuple (assume for now it's not just a single bool, though that should be allowed).
3. Consider the data structure as a tree: it has some branches (internal tree-shaped nodes) and some leaves (primitives, probably booleans). For each leaf that is not `true`, we want to report a failure and include the _path_ through the branches to that leaf, the comparison that failed, and what the expected and actual values were.

To do this, we first evaluate the file and get the data structure and walk through all of the nodes. For each leaf node, add an `Entry` to a list with the path we took to get there. If the leaf is not `true`, then we do some more work to complete the report.

I'm going to make a key assumption here because it simplifies things a lot and will be largely true in practice. Assume that the data structure containing the test output is statically defined (i.e. a literal, not generated by some other code). Each branch is either a set, array, dict or tuple, and each leaf is some other kind of expression.

So to get the failed comparison, parse the test file and walk the AST in search of the lowest tuple or sequence attribute found in the failed path. If there are multiple (e.g. `empty` which could be used in different cases), work your way back up through the parent nodes to find the right one.

Once you have the leaf node (unevaluated at this point), search inside that for a comparison operator (something that returns true: `=`, `>`, `<:`, `(<=)`, etc.). When you find one, replace it with some analogue to `//test.assert*` which, instead of returning `true` or `false`, evaluates to a tuple with the data you want to report (expected source and value, actual source and value).

Evaluate that modified AST and walk it, compiling the report from the leaf tuples. Fin.

I expect that I'm glossing over the complexity of "find the comparison", but it's conceptually possible.

You might argue "wow that's hard; this PR is an improvement, shouldn't we just merge it and do the more complete solution as a follow-up?" Maybe. I'm wary because a) it's mixing test code with implementation which is a code smell for both test and implementation code, and b) it's not in the same direction as the solution described above; we would ultimately plan to delete this test code, so why commit it in the first place and add an extra complexity and maintenance burden?